### DISCUSSION TITLE
Hands-on Discussion: Chapter 7 - Model specialization

### DISCUSSION THREAD

**Discussion by peremartra** [peremartra/Rearchitecting-LLMs#16] [General]
## Discussion

This is the space to share your findings from the experiments in Chapter 7. We moved from structural modification to behavioral specialization: QLoRA and QDoRA transferred a fifteen-line prompt into the model's weights, taking schema compliance from 85.0% (strict prompt) to 100.0% (single-word instruction), at the cost of a moderate drop in general capabilities (-3.2% weighted average, with ARC-Easy hit hardest).

The chapter's notebook (`CH07_NB02_L4_QLoRA_QDoRA.ipynb`) uses r=8 and 3 epochs. There's also a more aggressive variant with r=1 and a single epoch that keeps 100% compliance while preserving capabilities better. Share your numbers from either direction.

## 🎯 The Challenges

Pick one (or more) and share what you find:

1. **The data efficiency frontier**: Using r=8, 3 epochs, train on shrinking subsets of the training data, from 250 samples down to 30. Keep the 40-sample test set fixed. At what point does schema compliance collapse? Gradual or sharp threshold? Which category breaks first, irrelevant, typos, or implicit?

2. **Target module selection**: Train three QLoRA variants (r=8, 3 epochs), changing only `target_modules`: full Attention Module (`q_proj`, `k_proj`, `v_proj`, `o_proj`), full MLP Module (`gate_proj`, `up_proj`, `down_proj`), or the minimal original LoRA config (`q_proj`, `v_proj`). Does any reach 100% alone? Which module matters more, attention or MLP?

3. **Finding a task where DoRA wins**: Section 7.4 predicts DoRA's edge shows up on semantic adaptation, not structural reformatting. Swap the clinical dataset for a subset of `databricks-dolly-15k`, keeping the rest of the pipeline unchanged. Does the QLoRA/QDoRA tie break?

4. **Aggressive config vs standard**: Run the r=1, single-epoch variant against the r=8, 3-epoch baseline. How much capability retention improves, and where does the lighter adapter start to struggle?

📊 Share Your Results

```
Model analyzed:            [e.g., SmolLM2-1.7B-Instruct]
Challenge attempted:       [e.g., Data Efficiency Frontier]
Key config changes:        [e.g., r=8, target_modules=attention only]
Schema compliance:         [e.g., 100% overall, 71% on irrelevant with 60 samples]
Benchmark degradation:     [e.g., weighted avg -2.1%, arc_easy -6.4%]
Training time / hardware:  [e.g., 12 min on T4]
Key insight:                [What trade-off or surprise did you find?]
```

Failed runs and configs that made things worse are just as welcome as clean wins.
